sfge/src/test/java/com/salesforce/rules/UnnecessarilyExtensibleClassRuleTest.java

```diff
@@ -0,0 +1,188 @@
+import org.hamcrest.MatcherAssert;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class UnnecessarilyExtensibleClassRuleTest {
+    private GraphTraversalSource g;
+
+    @BeforeEach
+    public void setup() {
+        this.g = TestUtil.getGraph();
+    }
+
+    @MethodSource("paramProvider_testOuterClasses")
+    @ParameterizedTest(name = "{displayName}: Scope {0}")
+    public void testOuterClasses(String scope, Set<String> expectedDefiningTypes) {
+        // This class is never extended.
+        String unextendedSource = scope + " class UnextendedClass {}";
+        // This class is extended.
+        String extendedSource = scope + " class ExtendedClass {}";
+        // This class does the extension.
+        String extenderSource = "public class ExtenderClass extends ExtendedClass {}";
+
+        TestUtil.buildGraph(g, unextendedSource, extendedSource, extenderSource);
+
+        StaticRule rule = UnnecessarilyExtensibleClassRule.getInstance();
+        List<Violation> violations = rule.run(g);
+
+        MatcherAssert.assertThat(
+                "Wrong number of violations found",
+                violations,
+                hasSize(expectedDefiningTypes.size()));
+        for (int i = 0; i < expectedDefiningTypes.size(); i++) {
+            String actual = violations.get(i).getSourceDefiningType();
+            assertTrue(
+                    expectedDefiningTypes.contains(actual),
+                    String.format("%s should not be unextended", actual));
+        }
+    }
+
+    @MethodSource("paramProvider_testInnerClasses")
+    @ParameterizedTest(name = "{displayName}: Scope {0}")
+    public void testInnerClasses(String scope, Set<String> expectedDefiningTypes) {
+        String innerClassSource =
+                "global class HasInnerClasses {\n"
+                        + scope
+                        + " class InternallyUsedClass {}\n"
+                        + scope
+                        + " class ExternallyUsedClass {}\n"
+                        + scope
+                        + " class UnusedClass {}\n"
+                        + "public class InternalUserClass extends InternallyUsedClass {}\n"
+                        + "}\n";
+        String externalUserSource =
+                "public class ExternalUserClass extends HasInnerClasses.ExternallyUsedClass {}\n";
+        TestUtil.buildGraph(g, innerClassSource, externalUserSource);
+
+        StaticRule rule = UnnecessarilyExtensibleClassRule.getInstance();
+        List<Violation> violations = rule.run(g);
+
+        MatcherAssert.assertThat(
+                "Wrong number of violations found",
+                violations,
+                hasSize(expectedDefiningTypes.size()));
+        for (int i = 0; i < expectedDefiningTypes.size(); i++) {
```

Comment:
Would be helpful to have these checks extracted to helper methods so that the check continues to be readable.